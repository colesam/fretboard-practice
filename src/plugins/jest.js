// This file is simply a helper for IDE's to recognize the jest globals
export const describe = global.describe
export const it = global.it
export const expect = global.expect
export const beforeEach = global.beforeEach
export const fn = global.fn
