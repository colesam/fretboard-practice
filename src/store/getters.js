// noinspection JSCommentMatchesSignature
/**
 * Methods that use logic to return complex data from the vuex store
 * @namespace Vuex.Getters
 */
export default {
  /**
   * Gets a list of all available boards
   * @memberof Vuex.Getters
   * @param {Object} state
   */
  getBoardList: state => Object.keys(state.boards).map(id => state.boards[id])
}
