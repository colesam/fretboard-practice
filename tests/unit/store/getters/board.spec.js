import { describe, it, expect } from '@/plugins/jest' // useful for IDE's

import state from '@/store/testData/testState.js'
import getters from '@/store/getters.js'

let { getBoardList } = getters

// Need to get functions returned from parameterized getters
const boardIncludesPosition = getters.boardIncludesPosition(state)
const getNoteAtPosition = getters.getNoteAtPosition(state)
const getBoardNoteMatrix = getters.getBoardNoteMatrix(state, { getNoteAtPosition })

// Main testing board
const board = state.boards['1']

describe('getBoardList', () => {
  it('returns an array of all boards in the vuex store', () => {
    const result = getBoardList(state)
    expect(Array.isArray(result)).toBe(true)
    expect(result).toMatchSnapshot()
  })
})

describe('boardIncludesPosition', () => {
  it('returns true if the position is included in the board', () => {
    expect(boardIncludesPosition(board.id, { fret: 3, string: 0 })).toBe(true)
  })

  it('returns false if the position is not included in the board', () => {
    expect(boardIncludesPosition(board.id, { fret: 19, string: 3 })).toBe(false)
  })
})

describe('getNoteAtPosition', () => {
  it('returns the correct note at the passed position', () => {
    const { id } = board
    expect(getNoteAtPosition(id, { fret: 0, string: 0 })).toBe('E')
    expect(getNoteAtPosition(id, { fret: 19, string: 3 })).toBe('D')
    expect(getNoteAtPosition(id, { fret: 3, string: 2 })).toBe('F')
    expect(getNoteAtPosition(id, { fret: 4, string: 2 })).toBe('F#')
  })

  it('returns the correct note on boards with a startIndex greater than 0', () => {
    const id = '3'
    expect(getNoteAtPosition(id, { fret: 0, string: 0 })).toBe('B')
    expect(getNoteAtPosition(id, { fret: 1, string: 1 })).toBe('F')
    expect(getNoteAtPosition(id, { fret: 2, string: 2 })).toBe('B')
    expect(getNoteAtPosition(id, { fret: 3, string: 3 })).toBe('F')
  })
})

describe('getBoardNoteMatrix', () => {
  it('returns a matrix of every note at every position on the fretboard', () => {
    expect(getBoardNoteMatrix('2')).toMatchSnapshot()
    expect(getBoardNoteMatrix('3')).toMatchSnapshot()
  })
})
