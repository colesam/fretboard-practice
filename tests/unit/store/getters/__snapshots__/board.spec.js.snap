// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBoardList returns an array of all boards in the vuex store 1`] = `
Array [
  Board {
    "id": "1",
    "name": "G major arpeggio - pos. 1",
    "noteMatrix": null,
    "notePreferences": Object {
      "A": "A",
      "A#": "B♭",
      "B": "B",
      "C": "C",
      "C#": "C#",
      "D": "D",
      "D#": "E♭",
      "E": "E",
      "F": "F",
      "F#": "F#",
      "G": "G",
      "G#": "A♭",
    },
    "numFrets": 19,
    "numStrings": 4,
    "openTuning": Array [
      "E",
      "A",
      "D",
      "G",
    ],
    "positions": Array [
      Object {
        "fret": 3,
        "string": 0,
      },
      Object {
        "fret": 2,
        "string": 1,
      },
      Object {
        "fret": 5,
        "string": 1,
      },
      Object {
        "fret": 0,
        "string": 2,
      },
      Object {
        "fret": 0,
        "string": 3,
      },
      Object {
        "fret": 4,
        "string": 3,
      },
      Object {
        "fret": 5,
        "string": 2,
      },
    ],
    "root": "G",
    "startingFret": 0,
  },
  Board {
    "id": "2",
    "name": "C major arpeggio - pos. 1",
    "noteMatrix": null,
    "notePreferences": Object {
      "A": "A",
      "A#": "B♭",
      "B": "B",
      "C": "C",
      "C#": "C#",
      "D": "D",
      "D#": "E♭",
      "E": "E",
      "F": "F",
      "F#": "F#",
      "G": "G",
      "G#": "A♭",
    },
    "numFrets": 6,
    "numStrings": 4,
    "openTuning": Array [
      "E",
      "A",
      "D",
      "G",
    ],
    "positions": Array [
      Object {
        "fret": 5,
        "string": 3,
      },
      Object {
        "fret": 2,
        "string": 2,
      },
      Object {
        "fret": 5,
        "string": 2,
      },
      Object {
        "fret": 3,
        "string": 1,
      },
      Object {
        "fret": 0,
        "string": 0,
      },
      Object {
        "fret": 3,
        "string": 0,
      },
    ],
    "root": "C",
    "startingFret": 0,
  },
  Board {
    "id": "3",
    "name": "A major scale - pos. 2",
    "noteMatrix": null,
    "notePreferences": Object {
      "A": "A",
      "A#": "B♭",
      "B": "B",
      "C": "C",
      "C#": "C#",
      "D": "D",
      "D#": "E♭",
      "E": "E",
      "F": "F",
      "F#": "F#",
      "G": "G",
      "G#": "A♭",
    },
    "numFrets": 7,
    "numStrings": 4,
    "openTuning": Array [
      "E",
      "A",
      "D",
      "G",
    ],
    "positions": Array [
      Object {
        "fret": 0,
        "string": 0,
      },
      Object {
        "fret": 2,
        "string": 0,
      },
      Object {
        "fret": 3,
        "string": 0,
      },
      Object {
        "fret": 5,
        "string": 0,
      },
      Object {
        "fret": 0,
        "string": 1,
      },
      Object {
        "fret": 2,
        "string": 1,
      },
      Object {
        "fret": 4,
        "string": 1,
      },
      Object {
        "fret": 5,
        "string": 1,
      },
      Object {
        "fret": 0,
        "string": 2,
      },
      Object {
        "fret": 2,
        "string": 2,
      },
      Object {
        "fret": 4,
        "string": 2,
      },
      Object {
        "fret": 5,
        "string": 2,
      },
      Object {
        "fret": 0,
        "string": 3,
      },
      Object {
        "fret": 2,
        "string": 3,
      },
      Object {
        "fret": 4,
        "string": 3,
      },
      Object {
        "fret": 6,
        "string": 3,
      },
    ],
    "root": "A",
    "startingFret": 7,
  },
]
`;
